### DISCUSSION TITLE
Type narrowing for two variables

### DISCUSSION THREAD

**Discussion by NitishKumar525** [denoland/deno#25492] [Q&A] [answered]
```
 interface A {
   type: "A";
   propsOfA: string;
}

interface B {
   type: "B";
   propsOfB: number;
}

interface C {
   type: "C";
   propsOfC: boolean;
}

function isEqual(a: A | B | C, b: A | B | C) {
  if (a.type !== b.type) {
    return false;
  }

  switch (a.type) {
    case "A": {
      return a.propsOfA === b.propsOfA;
    }
    case "B": {
      return a.propsOfB === b.propsOfB;
    }
    case "C": {
      return a.propsOfC === b.propsOfC;
    }
  }
}
```

**Error on all return statements in the switch case, because type b is not narrowed along with a (here shown for the first case):**

```
 Property 'propsOfA' does not exist on type 'A | B | C'.   
  Property 'propsOfA' does not exist on type 'B'. [2339]

```

Is there a clean way to do this? The only solution I see is to have 9 type checks aka. a switch/if for `b` in each case of `a`.

**Comment by marvinhagemeister** [comment]
Support for narrowing the parent type based on a property is not well supported in TS, except for a few cases like `switch` statements etc.

The folks over at the TypeScript repository might have a more detailed answer on why that is. Related https://github.com/microsoft/TypeScript/issues/42384

**Comment by mdazfar2** [comment, marked as answer]
@NitishKumar525 The Typescript compiler does not track dependencies between different variables in this way, so it doesn't "remember" that a.type and b.type are equal, for instance. This has been variously requested as a feature, but e.g. [this issue](https://github.com/microsoft/TypeScript/issues/54252) shows the Typescript team don't currently consider it worth implementing.

If it was my code, I would probably write it this way:

```bash
 switch (a.type) {
    case "A": {
      return b.type === "A" && a.propsOfA === b.propsOfA;
    }
    case "B": {
      return b.type === "B" && a.propsOfB === b.propsOfB;
    }
    case "C": {
      return b.type === "C" && a.propsOfC === b.propsOfC;
    }
  }
```

That's three checks instead of nine. However, this might look inelegant since `a.type === b.type` is one check rather than three. An alternative approach is to construct a version of the union type where it's OK to write `b.propsOfA` without narrowing on `b.type === "A"`:

```bash
 interface A {
    type: "A";
    propsOfA: string;
    propsOfB?: never;
    propsOfC?: never;
}

interface B {
    type: "B";
    propsOfA?: never;
    propsOfB: number;
    propsOfC?: never;
}

interface C {
    type: "C";
    propsOfA?: never;
    propsOfB?: never;
    propsOfC: boolean;
}
```

With the types declared this way, your original code compiles with no errors. This is also actually a little safer than the original version, since it forbids objects like `{type: "A", propsOfA: "foo", propsOfC: true}` which are assignable to the original type `A`.

If the above is too much to write out manually, you can also construct these types programmatically:

```bash
/**
 * Makes a discriminated union type "safe" by requiring properties specific
 * to some branches to be absent in other branches.
 */
type SafeUnion<T> = Simplify<T> extends infer _T ? _T extends unknown ? Replace<{[K in AnyKeyOf<T>]?: never}, _T> : never : never

/**
 * Simplifies a type to make it more readable.
 */
type Simplify<T> = T extends unknown ? {[K in keyof T]: T[K]} : never

/**
 * Similar to an intersection `S & T`, except shared properties take their
 * types only from `T`.
 */
type Replace<S, T> = Simplify<T & Omit<S, keyof T>>

/**
 * The keys of any member of a union type.
 */
type AnyKeyOf<T> = T extends unknown ? keyof T : never
```

Then your code can be written like:

```
type ABC = SafeUnion<A | B | C>

function isEqual(a: ABC, b: ABC) {
    // ...
}
```
[playground link](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgILIN4Chm7ATwAcIAuZAIlXIG4ddCoB7QgZwHkZUyWwpQBzWgF8sWUJFiIUAIUx1kBYmXLSa8hs3YxpZEAFcAtgCNow0eOjwkyAMJzcCoqQo21Djaw42yRxowA2EHAgZliKKKjSdgC8yADKcDAQAKogwIwgADzoAD7Isnk2AHyiMHogCGDpIMjALACiAI56cP4AFHBkkTYANMhGXVEAlPa1MMgdAHThyACE0bFG004j2A7IUBBgelA18P4sELS4InQsAO7AYAgAFhNwy8Sr8sgIcIcUVGRr6xtbOzUHh4tOgFotJsCOKhjg4ROs3h8VORvi9cJttrtkECmJ5tMgwf0ITitNIYScXgiUORXCjfn8MYCiZovPiFoTITAbGTkHCeVhTgB6ABUQpwQuQAFk4ABrCAsLHIAAmdQQfAMoDgkEVyHK1UcxAoLESEHI-Xwf2awD4IH4yA80CqcuQLGICGAMGACDFCkYzsYBhQRigwVuTrAvpMWKMh3AtRqjDAN2g-WDFSTLEmYoFYSc8WNqWqmQAKkV8fFgAZCP53fhi6WIAAPSAgRXy0BJKDIAD6ReQAH5u73G83WzqQNKQIxzjUBwAlCBVqSZDAAbQA0nG0CB8GuIPgOHWALp93QQABu0CEfR7pdPF87d+gomFouQ4riFar7uATrg+pQ4bIAYMooFcQGMJsfxwIqcBGIEmZvtmMwfpW1YwLWJZlkOTYQC28rlBOU4zpg66brK+CMOMRaHmQRbroeQjII+UDPiK3oocA-hwJ2gHBHGEiHJUeoAAZxMgABkyBFsJfSNkghBgM6NzcRA2r2lAjrymAIEKEmVreuE8oZP45owEwBjIMJ0kIUKSG5vOi5IJkcR9JhsQoV+6HFhJyBsOqYDOX05GUVJRQlFgL7ekWSbIORRnjME5oBsYyYhX+uoZP+Nl2QaqDbru+4wHWWHIMOuGjgRk7Tv2sV7iFvbMVgQA)

**Reply by NitishKumar525** [reply]
Thank you @mdazfar2, I went looking for an existing issue but didn't find one in a few minutes of search.
